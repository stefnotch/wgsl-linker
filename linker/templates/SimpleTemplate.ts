import { Template } from "../ModuleRegistry.ts";
import { sliceReplace, sliceWords } from "../Slicer.ts";

export const simpleTemplate: Template = {
  name: "simple",
  apply: (src, extParams) => {
    const slices = sliceWords(src, extParams);
    return sliceReplace(src, slices);
  },
};
