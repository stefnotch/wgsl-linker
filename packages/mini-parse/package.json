{
  "name": "mini-parse",
  "version": "0.5.0-d1",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "pr": "printenv",
    "prepublishOnly": "run-s build",
    "build": "run-s build:main build:util build:minified",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:minified": "vite build --minify --emptyOutDir=false",
    "test": "vitest",
    "organize": "organize-imports-cli tsconfig.json",
    "format": "prettier-eslint \"src/**/*.ts\" --write",
    "lint": "eslint src"
  },
  "type": "module",
  "repository": "github:wgsl-tooling-wg/wgsl-linker",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./minified": {
      "import": "./dist/minified.js",
      "types": "./dist/index.d.ts"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^22.5.0",
    "@typescript-eslint/eslint-plugin": "^8.12.2",
    "@typescript-eslint/parser": "^8.12.2",
    "@vitest/ui": "^2.1.4",
    "berry-pretty": "^0.0.5",
    "eslint": "^9.13.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "typescript": "^5.6.3",
    "vite": "^5.4.10",
    "vite-plugin-dts": "^4.3.0",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.1.4"
  }
}