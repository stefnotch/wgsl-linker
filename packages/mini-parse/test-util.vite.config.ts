/// <reference types="vitest" />
import assert from "node:assert";
import { resolve } from "path";
import { LibraryOptions, defineConfig } from "vite";
import baseViteConfig from "./vite.config.js";

const config = baseViteConfig;
assert(config.build);
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-test-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/test-util/index.ts")];
lib.fileName = "testUtil";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);

// {
//   plugins: [
//     tsconfigPaths(),
//     dts(), // generate
//     // visualizer({ brotliSize: true, gzipSize: true }), // generate stats.html size report
//   ],
//   build: {
//     emptyOutDir: false,
//     lib: {
//       name: "mini-parse-testing",
//       formats: ["es"],
//       entry: [resolve(__dirname, "./src/test-util/index.ts")],
//       fileName: "testUtil"
//     },
//     rollupOptions: {
//       // make sure to externalize deps that shouldn't be bundled
//       // into your library
//       external: ['vitest', 'mini-parse'],
//     },
//     minify: false,
//     sourcemap: true,
//   },
// });
