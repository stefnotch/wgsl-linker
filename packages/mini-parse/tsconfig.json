{
  "extends": ["../../tsconfig.options.json"],
  "compilerOptions": {
    "lib": ["DOM", "DOM.Iterable", "ESNext", "ESNext.AsyncIterable"],
    "paths": {
      "mini-parse": ["./src"],
      "mini-parse/test-util": ["./src/test-util"]
    }
  },
  "include": ["src/**/*", "*.config.ts"],
  "exclude": ["node_modules"]
}
