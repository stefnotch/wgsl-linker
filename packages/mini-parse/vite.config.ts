/// <reference types="vitest" />
import { resolve } from "path";
import { defineConfig } from "vite";
import dts from "vite-plugin-dts";

export default defineConfig({
  plugins: [dts()],
  test: { setupFiles: "./src/test/TestSetup.ts" },
  build: {
    lib: {
      name: "mini-parse",
      entry: [resolve(__dirname, "src/index.ts")],
    },
    minify: false,
    sourcemap: true,
  },
});
