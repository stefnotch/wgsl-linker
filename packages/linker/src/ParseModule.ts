import { srcLog, SrcMap } from "mini-parse";
import {
  AbstractElem,
  AliasElem,
  ExportElem,
  ExtendsElem,
  FnElem,
  GlobalDirectiveElem,
  ModuleElem,
  StructElem,
  TreeImportElem,
  VarElem,
} from "./AbstractElems.js";
import { processConditionals } from "./Conditionals.js";
import { parseWgslD } from "./ParseWgslD.js";

/** module with exportable text fragments that are optionally transformed by a templating engine */
export interface TextModule {
  kind: "text";
  exports: TextExport[];
  fns: FnElem[];
  vars: VarElem[];
  structs: StructElem[];
  imports: (ExtendsElem | TreeImportElem)[];
  aliases: AliasElem[];
  globalDirectives: GlobalDirectiveElem[];

  modulePath: string; // full path to the module e.g "package/sub/foo", or "_root/sub/foo"

  /** original src for module */
  src: string;

  /** src code after processing #if conditionals  */
  preppedSrc: string;

  /** tracks changes through conditional processing for error reporting */
  srcMap: SrcMap;
}

/** an export elem annotated with the fn/struct to which it refers */
export interface TextExport extends ExportElem {
  ref: FnElem | StructElem;
}

export function preProcess(
  src: string,
  params: Record<string, any> = {},
): SrcMap {
  return processConditionals(src, params);
}

export function parseModule(
  src: string,
  naturalModulePath: string,
  params: Record<string, any> = {},
): TextModule {
  const srcMap = preProcess(src, params);

  const preppedSrc = srcMap.dest;
  const parsed = parseWgslD(preppedSrc, srcMap);
  const exports = findExports(parsed, srcMap);
  const fns = filterElems<FnElem>(parsed, "fn");
  const aliases = filterElems<AliasElem>(parsed, "alias");
  const globalDirectives = filterElems<GlobalDirectiveElem>(
    parsed,
    "globalDirective",
  );
  const imports = parsed.filter(
    e => e.kind === "extends" || e.kind === "treeImport",
  ) as (ExtendsElem | TreeImportElem)[];
  const structs = filterElems<StructElem>(parsed, "struct");
  const vars = filterElems<VarElem>(parsed, "var");
  const overridePath = filterElems<ModuleElem>(parsed, "module")[0]?.name;
  matchMergeImports(parsed, srcMap);

  const modulePath = overridePath ?? naturalModulePath;
  // dlog({ modulePath, overridePath });
  const kind = "text";
  return {
    ...{ kind, src, srcMap, preppedSrc, modulePath },
    ...{ exports, fns, structs, vars, imports },
    ...{ aliases, globalDirectives },
  };
}

export function filterElems<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"],
): T[] {
  return parsed.filter(e => e.kind === kind) as T[];
}

function findExports(parsed: AbstractElem[], srcMap: SrcMap): TextExport[] {
  const results: TextExport[] = [];
  const exports = findKind<ExportElem>(parsed, "export");

  exports.forEach(([elem, i]) => {
    let next: AbstractElem | undefined;
    do {
      next = parsed[++i];
    } while (next?.kind === "extends");
    if (elem.kind === "export") {
      if (next?.kind === "fn" || next?.kind === "struct") {
        results.push({ ...elem, ref: next });
      } else {
        srcLog(srcMap, elem.start, `#export what? (#export a fn or struct)`);
      }
    }
  });
  return results;
}

/** fill in extendsElem field of structs */
function matchMergeImports(parsed: AbstractElem[], srcMap: SrcMap): void {
  const extendsElems = findKind<ExtendsElem>(parsed, "extends");
  extendsElems.forEach(([extendsElem, i]) => {
    let next: AbstractElem | undefined;
    do {
      next = parsed[++i];
    } while (next?.kind === "extends" || next?.kind === "export");
    if (next?.kind === "struct") {
      next.extendsElems = next.extendsElems ?? [];
      next.extendsElems.push(extendsElem);
    } else {
      srcLog(srcMap, extendsElem.start, `#extends not followed by a struct`);
    }
  });
}

function findKind<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"],
): [T, number][] {
  return parsed.flatMap((elem, i) =>
    elem.kind === kind ? ([[elem, i]] as [T, number][]) : [],
  );
}
