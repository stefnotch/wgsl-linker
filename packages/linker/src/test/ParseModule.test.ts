import { _withBaseLogger } from "mini-parse";
import { logCatch } from "mini-parse/test-util";
import { expect, test } from "vitest";
import { parseModule, TextModule } from "../ParseModule.js";
import { simpleTemplate } from "../templates/SimpleTemplate.js";

test("simple fn export", () => {
  const src = `
    export
    fn one() -> i32 {
      return 1;
    }
  `;
  const module = testParseModule(src);
  expect(module.exports.length).toBe(1);
  expect(module).toMatchSnapshot();
});

test("simple fn import", () => {
  const src = `
    import bar/foo

    fn bar() { foo(); }
  `;
  const module = testParseModule(src);
  expect(module.imports.length).toBe(1);
  expect(module).toMatchSnapshot();
});

test.skip("match #extends", () => {
  const src = `
    // #extends Foo from pkg
    // #extends Bar from pkg
    struct Elem {
      sum: f32
    }
  `;
  const module = testParseModule(src);
  const merges = module.structs[0].extendsElems!;
  expect(merges[0].name).toBe("Foo");
  expect(merges[1].name).toBe("Bar");
});

test("read simple struct export", () => {
  const exportPrefix = `export`;
  const src = `
    struct Elem {
      sum: f32
    }
  `;
  const module = testParseModule(exportPrefix + "\n" + src);
  expect(module.exports.length).toBe(1);
  const firstExport = module.exports[0];
  expect(firstExport.ref.name).toBe("Elem");
});

test("read #module", () => {
  const src = `
    module my.module.com
    export fn foo() {}
  `;
  const textModule = testParseModule(src);
  expect(textModule.modulePath).toBe("my.module.com");
});

test.skip("simple #template preserves src map", () => {
  const src = `
    #template simple
    fn foo() { XX }
  `;
  const expected = `
    fn foo() { /**/ }
  `;
  const templates = new Map([["simple", simpleTemplate.apply]]);
  const textModule = parseModule(src, "./foo", { XX: "/**/" }, templates);
  expect(textModule.preppedSrc).toContain("fn foo() { /**/ }");
  expect(textModule.preppedSrc).toBe(expected);
  expect(textModule.srcMap.entries.length).toBe(3);
});

test.skip("parse error shows correct line after simple #template", () => {
  const src = `
    #template simple
    fn foo () { XX }
    fn () { } // oops
  `;
  const templates = new Map([["simple", simpleTemplate.apply]]);
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    parseModule(src, "./foo", { XX: "/**/" }, templates);
  });
  expect(logged()).toMatchInlineSnapshot(`
    "missing fn name
        fn () { } // oops   Ln 4
          ^"
  `);
});

test("parse error shows correct line after #ifdef ", () => {
  const src = `
    // #if FALSE
    foo
    bar
    // #endif
    fn () { } // oops
  `;
  const templates = new Map([["simple", simpleTemplate.apply]]);
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    parseModule(src, "./foo", { XX: "/**/" }, templates);
  });
  expect(logged()).toMatchInlineSnapshot(`
    "missing fn name
        fn () { } // oops   Ln 6
          ^"
  `);
});

test("parse error shows correct line after #ifdef and simple #template", () => {
  const src = `
    // #if FALSE
    foo
    bar
    // #endif
    // #template simple
    fn foo () { XX }
    fn () { } // oops
  `;
  const templates = new Map([["simple", simpleTemplate.apply]]);
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    parseModule(src, "./foo", { XX: "/**/" }, templates);
  });
  expect(logged()).toMatchInlineSnapshot(`
    "missing fn name
        fn () { } // oops   Ln 8
          ^"
  `);
});

test("import gleam style", () => {
  const src = `
    import my/foo

    fn bar() { foo(); }
  `;
  const module = testParseModule(src);
  expect(module.imports).toMatchSnapshot();
});

function testParseModule(src: string): TextModule {
  return parseModule(src, "./test.wgsl");
}
