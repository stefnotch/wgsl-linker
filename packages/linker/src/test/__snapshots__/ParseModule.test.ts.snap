// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import gleam style 1`] = `
[
  {
    "end": 19,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "my",
        },
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "foo",
        },
      ],
    },
    "kind": "treeImport",
    "start": 5,
  },
]
`;

exports[`simple fn export 1`] = `
{
  "aliases": [],
  "exports": [
    {
      "end": 12,
      "kind": "export",
      "ref": {
        "calls": [],
        "end": 55,
        "kind": "fn",
        "name": "one",
        "nameElem": {
          "end": 22,
          "kind": "fnName",
          "name": "one",
          "start": 19,
        },
        "start": 16,
        "typeRefs": [
          {
            "end": 31,
            "kind": "typeRef",
            "name": "i32",
            "start": 28,
          },
        ],
      },
      "start": 5,
    },
  ],
  "fns": [
    {
      "calls": [],
      "end": 55,
      "kind": "fn",
      "name": "one",
      "nameElem": {
        "end": 22,
        "kind": "fnName",
        "name": "one",
        "start": 19,
      },
      "start": 16,
      "typeRefs": [
        {
          "end": 31,
          "kind": "typeRef",
          "name": "i32",
          "start": 28,
        },
      ],
    },
  ],
  "globalDirectives": [],
  "imports": [],
  "kind": "text",
  "modulePath": "./test.wgsl",
  "preppedSrc": "
    export
    fn one() -> i32 {
      return 1;
    }
  ",
  "src": "
    export
    fn one() -> i32 {
      return 1;
    }
  ",
  "srcMap": SrcMap {
    "dest": "
    export
    fn one() -> i32 {
      return 1;
    }
  ",
    "entries": [
      {
        "destEnd": 58,
        "destStart": 0,
        "src": "
    export
    fn one() -> i32 {
      return 1;
    }
  ",
        "srcEnd": 58,
        "srcStart": 0,
      },
    ],
  },
  "structs": [],
  "template": undefined,
  "vars": [],
}
`;

exports[`simple fn import 1`] = `
{
  "aliases": [],
  "exports": [],
  "fns": [
    {
      "calls": [
        {
          "end": 39,
          "kind": "call",
          "name": "foo",
          "start": 36,
        },
      ],
      "end": 44,
      "kind": "fn",
      "name": "bar",
      "nameElem": {
        "end": 31,
        "kind": "fnName",
        "name": "bar",
        "start": 28,
      },
      "start": 25,
      "typeRefs": [],
    },
  ],
  "globalDirectives": [],
  "imports": [
    {
      "end": 20,
      "imports": ImportTree {
        "segments": [
          SimpleSegment {
            "args": undefined,
            "as": undefined,
            "name": "bar",
          },
          SimpleSegment {
            "args": undefined,
            "as": undefined,
            "name": "foo",
          },
        ],
      },
      "kind": "treeImport",
      "start": 5,
    },
  ],
  "kind": "text",
  "modulePath": "./test.wgsl",
  "preppedSrc": "
    import bar/foo

    fn bar() { foo(); }
  ",
  "src": "
    import bar/foo

    fn bar() { foo(); }
  ",
  "srcMap": SrcMap {
    "dest": "
    import bar/foo

    fn bar() { foo(); }
  ",
    "entries": [
      {
        "destEnd": 47,
        "destStart": 0,
        "src": "
    import bar/foo

    fn bar() { foo(); }
  ",
        "srcEnd": 47,
        "srcStart": 0,
      },
    ],
  },
  "structs": [],
  "template": undefined,
  "vars": [],
}
`;
